WIP: Document the design of Nix (#4280)
NixOS/nix: doc/manual/src/design/overview.md

# Overview

> Nix is broken into layers that operate fairly independently.

Reviewer: ```suggestion
Nix consists of layers that operate fairly independently.
```
"Who broke it?"

> At the top is the *command line interface*, i.e. the argument parsing of the various Nix executables.

Author: I think there is no comma after i.e.? I would keep "the" as I had meant this to mean the concrete CLI, as opposed to the reusable stuff "libcmd".

> At the top is the *command line interface*, i.e. the argument parsing of the various Nix executables.

Reviewer: AE: ", i.e., ..."
BE: ", i.e ..."
https://english.stackexchange.com/a/6896

> At the top is the *command line interface*, i.e. the argument parsing of the various Nix executables.

Author: Alright! American native speaker learning he was doing the British one all along haha.

Below that is the Nix *expression language*, in which packages and configurations are written.
These are the layers which users interact with most.

> Below that is the *store layer*, Nix's machinery for presenting and files and fully elaborated build plans, and also executing those build plans.

Author: Remember at this point "derivation" has not been introduce yet :). but I think everything before that is good.

> Below that is the *store layer*, Nix's machinery for presenting and files and fully elaborated build plans, and also executing those build plans.

Author: ```suggestion
Below that is the *store layer*, Nix' machinery to represent tracked files, dependencies, and fully elaborated build plans.
It is also used for executing those build plans.
```
Does that look good? Note I have been doing one sentence per line so the diffs stay readible.

The store layer may not be as visible, but this is the heart of Nix.

> This chapter will start there and work up towards the more user-facing interfaces described in the rest of the manual.

Reviewer 2: My first reading of this sentence left me confused as to what "there" is referring to. I would find it easier to understand as follows:

"This chapter starts with the store layer and works towards the user-facing interfaces described in the rest of the manual."

I also think this is implying a statement of scope, but I don't quite understand what the scope is supposed to be. Perhaps the following is correct: 

"This chapter describes the store layer of nix and how it interacts with the expression language and the command line interfaces described in the other chapters in this manual."

Author: ```suggestion
This chapter describes Nix starting with that bottom store layer, then working its way up until it reaches the more user-facing interfaces described in the rest of the manual."
```

Author: Does this take into account your points?

> This chapter will start there and work up towards the more user-facing interfaces described in the rest of the manual.

Author: The "overall. unabridged" plan to still talk about the later layers too. We can PR them separately as discussed, so at PR time would we find a different way to phrase this.

> This chapter describes Nix starting with that bottom store layer, then working its way up until it reaches the more user-facing interfaces described in the rest of the manual."

Author: Wait narrow scope in later PRs.
